### ISSUE TITLE
Docs for using broadcast address and rpc bind address

### ISSUE THREAD

**Issue by rkarthik007** [yugabyte/yugabyte-db#390]
Need docs for:
[#389] Support for public ip address, dns names and private ip address

**Comment by kriteshranjan** [comment]
Closing this issue based on AI validation. cc: @m-iancu , @VamsidharGunturu

**Comment by kriteshranjan** [comment]
AI validation notes (`yb-issue-390-analysis.md`):

# Analysis: yugabyte/yugabyte-db#390 (no Jira key visible; `temp/docs-migration` label = Jira-migration marker)

## Verdict
NOT-A-BUG / already-resolved (app verdict: **Fixed**) — the requested documentation exists in the current docs tree and covers the full ask; the issue was simply never closed.

## The ask
Issue #390 ("Docs for using broadcast address and rpc bind address", filed 2018-07-20 by rkarthik007, area/documentation, 0 comments, OPEN) body is one line: "Need docs for: [#389] Support for public ip address, dns names and private ip address". #389 (closed 2019-04-04, implemented by `e65b6ed72d`) specifies exactly what needed documenting:

1. `rpc_bind_addresses` — interfaces to bind for RPC; typically private IP; `0.0.0.0` allowed.
2. `server_broadcast_addresses` — externally specified public IP / DNS name (server may not know its own), optional port.
3. `use_private_ip` — inter-node communication policy: `never` (default) / `cloud` / `region` / `zone`.
4. "Listen Address" — explicitly "a future task" in #389; never shipped as a separate flag → outside the doc ask.

## Current state (root-cause equivalent for a docs issue)
Master checkout `b89bc62335` (reviewed 2026-08-16), docs tree `docs/content/stable/`:

- `reference/configuration/yb-tserver.md:717-719` — "Networking" / "RPC and binding addresses" section.
- `reference/configuration/yb-tserver.md:734-755` — `--rpc_bind_addresses`: semantics, private-IP default, and the when-to-set-broadcast warning at :752-755 ("Make sure that the server_broadcast_addresses flag is set correctly if... rpc_bind_addresses is set to 0.0.0.0 ... involves public IP addresses").
- `reference/configuration/yb-tserver.md:757-764` — `--server_broadcast_addresses`: "public IP or DNS hostname of the server (with an optional port). This value is used by servers to communicate with one another, depending on the connection policy parameter."
- `reference/configuration/yb-tserver.md:779-793` — `--use_private_ip`: full `never`/`zone`/`cloud`/`region` policy table matching #389's spec verbatim, cross-linked to geo-distribution flags.
- `reference/configuration/yb-master.md:332-353, 355-362, 382-396` — same three flags for yb-master.
- `reference/configuration/default-ports.md:40` — cross-links all three flag anchors.
- `yugabyte-voyager/reference/data-migration/import-data.md:172` and `bulk-data-load/import-data-file.md:166` — note configuring `server-broadcast-addresses` for public-IP clusters.
- Deploy guides (`deploy/manual-deployment/start-masters.md:66-161`, `deploy/multi-dc/3dc-deployment.md:53-89`) instruct private-IP `rpc_bind_addresses` and link to the reference entries; `deploy/checklist.md:321-341` covers ports/network access. No dedicated NAT/public-IP walkthrough page exists, but that tutorial was delivered as the 1.1 blog post rkarthik007 linked from #389 (2018-10-20), and the reference docs carry the complete flag semantics + when-to-set guidance the issue asked for.

History: docs mentioning `server_broadcast_addresses`/`use_private_ip` first landed `16a54fa024` (2018-10-01, v1.1 docs) — ~10 weeks after #390 was filed. Same content persists in versioned trees (v2.20, v2024.2, v2025.1, v2025.2).

## Coverage matrix (predicted vs observed)
| Ask item (#389 spec) | Predicted (from #326 memory) | Observed |
|---|---|---|
| `rpc_bind_addresses` semantics | documented | yb-tserver.md:734-755; yb-master.md:332-353 — covered |
| `server_broadcast_addresses` (public IP/DNS, optional port) | documented | yb-tserver.md:757-764; yb-master.md:355-362 — covered |
| When to set broadcast (0.0.0.0 / public-IP bind) | documented (memory cited 752-762) | yb-tserver.md:752-755; yb-master.md:350-353 — covered |
| `use_private_ip` policy never/zone/cloud/region | unknown | yb-tserver.md:779-793; yb-master.md:382-396 — covered, matches #389 enum |
| Listen address | out of scope ("future task" per #389) | correctly absent as separate flag |
| Task-oriented deploy walkthrough (NAT/public IP) | partial | deploy guides link to reference; tutorial lived in 1.1 blog post — reference coverage satisfies the ask |

## Reproduction
N/A — documentation request, no code defect. Rerunnable evidence:

```sh
gh issue view 390 --repo yugabyte/yugabyte-db --json title,body,labels,state
gh issue view 389 --repo yugabyte/yugabyte-db --json title,state,closedAt
rg -n "rpc_bind_addresses|server_broadcast_addresses|use_private_ip" \
  docs/content/stable/reference/configuration/yb-tserver.md \
  docs/content/stable/reference/configuration/yb-master.md
git log --oneline --reverse -S server_broadcast_addresses -- docs/ | head -1  # 16a54fa024 (2018-10-01)
```

## Versions tested
Source review only (no builds/containers, per docs-issue nature): master `b89bc62335` docs tree, reviewed 2026-08-16. Master affected: **no** — requested docs present.

## temp/docs-migration tracking
Label added 2023-02-04 by polarweasel ("TEMPORARY USE for migrating docs tickets to Jira") — bulk docs-team migration to internal Jira. No public GitHub docs-migration epic cross-references #390 (timeline shows only the two label events + two unrelated `referenced` events); no Jira key visible on the issue. Label cohort still holds many open GitHub issues (e.g. #15955, #15764), i.e. GitHub side was not closed post-migration.

## Proposed fix direction
No code or docs change needed. Recommend closing #390 as completed, citing:
- `docs/content/stable/reference/configuration/yb-tserver.md` §§ `--rpc_bind_addresses` / `--server_broadcast_addresses` / `--use_private_ip`
- `docs/content/stable/reference/configuration/yb-master.md` same sections
Optional polish (not required by the ask): a task-oriented public-IP/NAT deployment page under `deploy/`, replacing the aging blog-post link.

## Test strategy
N/A — documentation issue; validation is the coverage review above.

## Confidentiality check
All sources public (GitHub issues #390/#389, public docs tree, public git history). No customer data, PII, internal hostnames, or secrets.

## Related
- #326 validated 2026-08-16 (same networking machinery; artifact /tmp/claude/yb-issue-326-analysis.md); code fix history: `668da23103` (2018-07-16) introduced `server_broadcast_addresses`/`use_private_ip`; `e65b6ed72d` (2018-08-22) completed #389.

**Comment by kriteshranjan** [comment]
AI validation notes (`yb-issue-326-analysis.md`):

# Analysis: yugabyte/yugabyte-db#326 (Jira DB-4716)

## Verdict
NOT-A-BUG on current master (app verdict: **Fixed**) — the 2018 wildcard-bind
inter-node-connectivity defect class was resolved by the broadcast/advertised-address
machinery landed after filing; both resolution branches the issue offered
("figure out root cause" and "improve our docs") are done.

## Issue summary
Filed 2018-06-12 by a maintainer (state OPEN, 0 comments). Body: per discussion in
#318, "we sometimes have issues with node to node connectivity when RPC bind
addresses are set to `0.0.0.0`"; @spolitov suspected "some ipv6 related issue".
Asks: fix root cause, or document correct flag setup. Kind: bug (connectivity),
component DocDB (`src/yb/server`, `src/yb/util/net`, `src/yb/common`).

## Root cause (historical) and current state
At filing (tree of 2018-06-12, e.g. `49c9d4a1bd`), `AddHostPortPBs()` already
substituted wildcard-bound endpoints with `GetFQDN()` (fallback: unordered
`GetLocalAddresses(EXTERNAL)`), but:

1. **No override existed.** If the FQDN resolved to something peers cannot reach
   (loopback via `/etc/hosts`, an IPv6/link-local address, a NAT-private IP),
   the advertised address was wrong and there was no knob to correct it.
2. **No IPv4/IPv6 preference ordering** in local-address selection — matching
   spolitov's IPv6 suspicion.

Current master (`b89bc62335`, 2026-08-14) — mechanism, wildcard never advertised verbatim:

- `src/yb/server/rpc_server.cc:58` — `rpc_bind_addresses` defaults to `0.0.0.0`;
  wildcard bind is the shipped default, exercised by every default deployment.
- `src/yb/server/server_base.cc:594-607` — `RpcAndWebServerBase::GetRegistration()`
  detects wildcard input (`IsWildcardAddress`, `src/yb/util/net/net_util.cc:669-673`)
  and falls back to bound endpoints passed through `AddHostPortPBs`.
- `src/yb/common/wire_protocol.cc:350-382` — `AddHostPortPBs()`: unspecified
  endpoint → advertise `GetFQDN()` (`net_util.cc:431-461`); on failure, enumerate
  concrete local IPs via `GetLocalAddresses(FLAGS_net_address_filter, ...)`.
- `src/yb/util/net/net_util.cc:80-87` — `net_address_filter` default
  `ipv4_external,ipv4_all,ipv6_external,ipv6_non_link_local,ipv6_all` prefers
  external IPv4 first (directly addresses the IPv6 theory).
- `src/yb/server/server_base_options.cc:94-108,161` + `server_base.cc:609` —
  `--server_broadcast_addresses` parsed and registered in `ServerRegistrationPB`;
  the explicit operator override when auto-derived addresses are wrong.
- `src/yb/common/wire_protocol.cc:405-452` — consumers pick broadcast vs private
  per `use_private_ip` policy (`DesiredHostPort`/`UsePublicIp`; flag at :56).
- `src/yb/util/net/net_util.cc:477-496` — `HostPortFromEndpointReplaceWildcard()`
  same substitution for single endpoints; unit-tested at
  `src/yb/util/net/net_util-test.cc:190-193` (`EXPECT_NE("0.0.0.0", hp.host())`).

Docs branch: `docs/content/stable/reference/configuration/yb-tserver.md:752-762`
explicitly instructs setting `server_broadcast_addresses` when
`rpc_bind_addresses` is `0.0.0.0` (yb-master.md analog exists). #390 (docs
umbrella) remains OPEN but the reference docs cover the filed pitfall.

## Fixing/hardening commits
- `668da23103` (2018-07-16, D5054, ENG-3279) — public/private IP support:
  `server_broadcast_addresses`, `use_private_ip`, `DesiredHostPort` selection.
  Landed ~5 weeks after filing; same era/author group as the issue discussion.
- `5a88e2fe05` (2020-06-02, #3644) — `net_address_filter` IPv4/IPv6 preference
  ordering for bind/advertised address selection.
- `8f906a215c` (2020-08-27, #2540) — default port for `server_broadcast_addresses`.
- `1b4987bf9a` (2020-09-15, #5320) — webserver_interface defaults from rpc bind.
- `08d5ccbbc9` (2020-10-21, #5641) — TLS cert lookup fixed when bound to `0.0.0.0`.
- No commit references GH #326 directly (`git log --all --grep '#326\b'` hits only
  vendored third-party).

## Trigger matrix (code-derived; predicted == observed via source/doc/commit evidence)
| condition | 2018 behavior (filed) | master b89bc62335 |
|---|---|---|
| bind 0.0.0.0, hostname/FQDN peer-resolvable | works | works — FQDN advertised (wire_protocol.cc:357) |
| bind 0.0.0.0, FQDN resolves to loopback/unreachable | broken, no remedy | remediable: `--server_broadcast_addresses` advertised to peers (server_base.cc:609; wire_protocol.cc:431-452) |
| bind 0.0.0.0, local hostname lookup fails | unordered EXTERNAL local addrs (v6 possible) | ordered by `net_address_filter`, external IPv4 first (net_util.cc:80-87) |
| wildcard advertised verbatim to peers | never (substitution existed) | never — guarded + unit-tested (net_util-test.cc:191) |

## Reproduction
N/A — obsolete-report validation; no repro attempted (fix-class landed 2018-2020,
wildcard bind is today's default and default multi-node deployments work).
Rerunnable evidence commands:
```sh
gh issue view 326 --repo yugabyte/yugabyte-db --json title,body,state,comments
git log -1 --format='%h %ci'                       # b89bc62335 2026-08-14
rg -n 'rpc_bind_addresses' src/yb/server/rpc_server.cc          # :58 default 0.0.0.0
rg -n 'IsWildcardAddress|AddHostPortPBs' src/yb/server/server_base.cc  # :599,:603
sed -n '350,382p' src/yb/common/wire_protocol.cc   # wildcard -> FQDN/local-IP substitution
git show --stat 668da23103 | head -30              # ENG-3279 broadcast machinery
rg -n 'server_broadcast_addresses' docs/content/stable/reference/configuration/yb-tserver.md
```

## Versions tested
Source review only: master `b89bc62335` (2026-08-14); historical tree `49c9d4a1bd`
(2018-03-30, last wire_protocol.cc change before filing). Master affected: **no**.

## Proposed fix direction
None (code). Housekeeping: issue can be closed as fixed-by `668da23103` +
follow-ups, with docs pointer to yb-tserver.md `--server_broadcast_addresses`
section; residual environmental pitfall (advertised FQDN must be peer-resolvable)
is a docs matter tracked by still-open #390.

## Test strategy
Existing coverage: `src/yb/util/net/net_util-test.cc:185-199`
(`NetUtilTest.TestReverseLookup` asserts wildcard is never advertised).
Public/private selection covered by MiniCluster tests exercising
`TEST_check_broadcast_address` connectivity simulation (`server_base.cc:789-810`).
No new tests needed.

## Confidentiality check
Public-source facts only: public GitHub issues (#326, #318, #389, #390), public
repo code/doc citations, public commit hashes. No customer data, PII, internal
hostnames, or secrets.
